hack/generator/pkg/astmodel/armconversion/from_arm_conversion_function.go

```diff
@@ -0,0 +1,431 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package armconversion
+
+import (
+	"fmt"
+	"github.com/Azure/k8s-infra/hack/generator/pkg/astbuilder"
+	"github.com/Azure/k8s-infra/hack/generator/pkg/astmodel"
+	"go/ast"
+	"go/token"
+)
+
+func (c *ArmConversionFunction) asConvertFromArmFunc(
+	codeGenerationContext *astmodel.CodeGenerationContext,
+	receiver astmodel.TypeName,
+	methodName string) *ast.FuncDecl {
+
+	// Determine the type we're operating on
+	kubeType := getReceiverObjectType(codeGenerationContext, receiver)
+
+	var fromArmStatements []ast.Stmt
+
+	receiverIdent := ast.NewIdent(c.idFactory.CreateIdentifier(receiver.Name(), astmodel.NotExported))
+
+	// Note: If we have a property with these names we will have a compilation issue in the generated
+	// code. Right now that doesn't seem to be the case anywhere but if it does happen we may need
+	// to harden this logic some to choose an unused name.
+	inputIdent := ast.NewIdent("armInput")
+	typedInputIdent := ast.NewIdent("typedInput")
+
+	// perform a type assert
+	fromArmStatements = append(fromArmStatements, &ast.AssignStmt{
+		Lhs: []ast.Expr{
+			typedInputIdent,
+			ast.NewIdent("ok"),
+		},
+		Tok: token.DEFINE,
+		Rhs: []ast.Expr{
+			&ast.TypeAssertExpr{
+				X:    inputIdent,
+				Type: ast.NewIdent(c.armTypeName.Name()),
+			},
+		},
+	})
+
+	// Check the result of the type assert
+	fromArmStatements = append(fromArmStatements, &ast.IfStmt{
+		Cond: &ast.UnaryExpr{
+			Op: token.NOT,
+			X:  ast.NewIdent("ok"),
+		},
+		Body: &ast.BlockStmt{
+			List: []ast.Stmt{
+				&ast.ReturnStmt{
+					Results: []ast.Expr{
+						&ast.CallExpr{
+							Fun: &ast.SelectorExpr{
+								X:   ast.NewIdent("fmt"),
+								Sel: ast.NewIdent("Errorf"),
+							},
+							Args: []ast.Expr{
+								&ast.BasicLit{
+									Kind:  token.STRING,
+									Value: fmt.Sprintf("\"unexpected type supplied for FromArm function. Expected %s, got %%T\"", c.armTypeName.Name()),
+								},
+								inputIdent,
+							},
+						},
+					},
+				},
+			},
+		},
```

Comment:
This is pretty ugly but not sure if there's a nicer way to do it... Can you shove it off into another function so you don't have to stare at it and it can have a helpful name describing what this is actually doing?

Reply:
Yeah, basically all of the "using the ast" code ends up really ugly -- I created a helper for some common stuff, possibly I can turn this into one too.

Reply:
Did this, it's somewhat better now.